lib/rules/object-shorthand.js

```diff
@@ -284,18 +284,21 @@ module.exports = {
             const arrowToken = sourceCode.getTokenBefore(node.value.body, astUtils.isArrowToken);
             const fnBody = sourceCode.text.slice(arrowToken.range[1], node.value.range[1]);
 
-            let shouldAddParensAroundParameters = false;
             let tokenBeforeParams;
+            let shouldAddParensAroundParameters = false;
 
-            if (node.value.params.length === 0) {
-                tokenBeforeParams = sourceCode.getFirstToken(node.value, astUtils.isOpeningParenToken);
+            // First token should not be `async`
+            if (node.value.async) {
+                tokenBeforeParams = sourceCode.getFirstToken(node.value, token => !(token.type === "Identifier" && token.value === "async"));
             } else {
-                tokenBeforeParams = sourceCode.getTokenBefore(node.value.params[0]);
+                tokenBeforeParams = sourceCode.getFirstToken(node.value);
             }
```

Comment:
I think this works correctly, but is a bit difficult to follow as `tokenBeforeParams` can be the first token before params, but also the first token of `node.value` that isn't the first token before params, or the first token of params, or even the arrow token (e.g., `async async => {}`) depending on the code encountered.

Could we maybe simplify the code this way:

```js
// First token should not be `async`
const firstValueToken = sourceCode.getFirstToken(node.value, { skip: node.value.async ? 1 : 0 });

const sliceStart = firstValueToken.range[0];
const sliceEnd = sourceCode.getTokenBefore(arrowToken).range[1];

const oldParamText = sourceCode.text.slice(sliceStart, sliceEnd);
const newParamText = node.value.params.length === 1 && node.value.params[0].range[0] === sliceStart
    ? `(${oldParamText})`
    : oldParamText;

return fixer.replaceTextRange(
    fixRange,
    methodPrefix + newParamText + fnBody
);
```

Reply:
Very nice, thanks!